### PR TITLE
[Linux] Add test logic to verify commissioning between Linux tv-casting-app and Linux tv-app.

### LINKED ISSUE #33300: [Linux] Add logic to test for commissioning between the Linux tv-casting-app and the Linux tv-app.
Continuing on from #32787 and #32918.

There is no CI check to test that the commissioning between the Linux tv-casting-app and the Linux tv-app continues to work whenever a PR is created. Therefore, this issue will add the logic needed to the `run_tv_casting_test.py` file to test that commissioning continues to work between the Linux tv-casting-app and the Linux tv-app. The output of both the tv-casting-app and tv-app will be parsed for strings of interest as part of the verification that commissioning is successful.

In order for commissioning to be considered successful, the following need to be true:
- We should be able to initiate the cast request by sending `cast request {valid_discovered_commissioner_number}` to the Linux tv-casting-app process, i.e. `cast request 0`.
- The `device Name`, `vendor id`, and `product id` in the `Identification Declaration` block from the Linux tv-casting-app output need to match the corresponding values noted in the `Identification Declaration` block from the Linux tv-app output.
- We should be able to approve the cast request command from the tv-casting-app to the tv-app by sending `controller ux ok` to the Linux tv-app process.
- We should be able to see `Commissioning completed successfully` in the tv-casting-app output and `commissioning success` on the tv-app output.

In the case of failure (when any of the above stages fail), we should print an error message, dump the logs to console, and exit on error.

### PR DESCRIPTION
**Pull request by shaoltan-amazon**
Fixes #33300 

**Problem**
There is no CI check to test that the commissioning between the Linux tv-casting-app and the Linux tv-app continues to work whenever a PR is created.

**Solution Overview**
Add testing logic to the `run_tv_casting_test.py` script to verify that commissioning continues to work between the tv-casting-app and the tv-app.

The output of the tv-casting-app and the tv-app will be parsed for strings of interest as they are received and the output will also be written to their corresponding log files.

The following has been added to the test script:
- The user (in this case the test script) will pass in `cast request {valid_discovered_commissioner_number}` to the tv-casting-app subprocess once we find the `cast request 0` string in the tv-casting-app output. This is the indication from the tv-casting-app to the user that it is ready for the cast request to be initiated.
- After the cast request is initiated by the user (in this case the test script), we will verify that the `device Name`, `vendor id`, and `product id` noted under the `Identification Declaration` block in the tv-casting-app output matches the corresponding values on the tv-app side.
- Once the device information is confirmed, we will parse the tv-app output for the `Via Shell Enter: controller ux ok|cancel` string. This will indicate that the user can approve the cast request by sending `controller ux ok` to the tv-app subprocess.
- Lastly, we will parse the output of both the tv-casting-app and the tv-app for messages indicating that commissioning was successful. These messages are `Commissioning completed successfully` and `PROMPT USER: commissioning success` respectively.
- If we're not able to find the strings of interest within the timeout that was set for each of the stages above or if we see `Commissioning failed` in the tv-casting-app output, we will print an error message, dump the logs to console, and exit on error.

**Testing**
Tested the script locally by running `python3 ./scripts/tests/run_tv_casting_test.py` and observing the output. Also verified that the updated test script works as expected in the CI check.

### CODE REVIEW

File: scripts/tests/run_tv_casting_test.py [hunk 1]
```diff
@@ -84,30 +88,35 @@ def handle_discovery_failure(log_file_paths: List[str]):
     sys.exit(1)
 
 
-def extract_value_from_string(line: str) -> int:
-    """Extract and return integer value from given output string.
+def extract_value_from_string(line: str) -> Union[int, str]:
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py:91]: Good suggestion, I will make the change.

File: scripts/tests/run_tv_casting_test.py [hunk 2]
```diff
@@ -134,7 +143,170 @@ def start_up_tv_app_success(tv_app_process: subprocess.Popen, linux_tv_app_log_f
...
+            product_id = extract_value_from_string(line)
+
+        if device_name and vendor_id and product_id:
+            break
+
+    return device_name, vendor_id, product_id
+
+
+def validate_tv_app_device_info(tv_app_info, device_name, vendor_id, product_id, log_paths) -> bool:
+    """Validate device information from the 'Identification Declaration' block from the Linux tv-app output against the expected values."""
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py:201]: Sure, I will update.

File: scripts/tests/run_tv_casting_test.py [hunk 3]
```diff
@@ -134,7 +143,170 @@ def start_up_tv_app_success(tv_app_process: subprocess.Popen, linux_tv_app_log_f
...
+    """Initiate commissioning between Linux tv-casting-app and tv-app by sending `cast request {valid_discovered_commissioner_number}` via Linux tv-casting-app process."""
+    tv_casting_app_process, linux_tv_casting_app_log_file = tv_casting_app_info
+
+    start_wait_time = time.time()
+
+    while True:
+        # Check if we exceeded the maximum wait time for initiating 'cast request' from the Linux tv-casting-app to the Linux tv-app.
+        if time.time() - start_wait_time > COMMISSIONING_STAGE_MAX_WAIT_SEC:
+            logging.error('The command `cast request ' + valid_discovered_commissioner_number +
+                          '` was not sent to the Linux tv-casting-app process within the timeout.')
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py:156]: I will update.

File: scripts/tests/run_tv_casting_test.py [hunk 4]
```diff
@@ -71,9 +75,9 @@ def dump_temporary_logs_to_console(log_file_path: str):
             print(line.rstrip())
 
 
-def handle_discovery_failure(log_file_paths: List[str]):
-    """Log 'Discovery failed!' as error, dump log files to console, exit on error."""
-    logging.error('Discovery failed!')
+def handle_casting_state_failure(casting_state: str, log_file_paths: List[str]):
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py:78]: I will update.

File: scripts/tests/run_tv_casting_test.py [hunk 5]
```diff
@@ -177,12 +349,33 @@ def parse_output_for_valid_commissioner(tv_casting_app_info: Tuple[subprocess.Po
...
+    # Extract the values from the 'Identification Declaration' block in the tv-casting-app output that we want to validate against.
+    device_name, vendor_id, product_id = extract_device_info(tv_casting_app_info)
+
+    if not validate_tv_app_device_info(tv_app_info, device_name, vendor_id, product_id, log_paths):
+        handle_casting_state_failure('Commissioning', log_paths)
+
+    if not approve_tv_casting_request(tv_app_info, log_paths):
+        handle_casting_state_failure('Commissioning', log_paths)
+
+    if not validate_commissioning_success(tv_casting_app_info, tv_app_info, log_paths):
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py:370]: Good idea, I will make the change.

**Review by sharadb-amazon — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**